subsys/net/lib/tftp/tftp_client.c

```diff
@@ -53,6 +53,61 @@ static size_t make_request(uint8_t *buf, int request,
 	return ptr - (char *)buf;
 }
 
+/*
+ * Send an Error Message to the TFTP Server.
+ */
+static int send_data(int sock, uint32_t block_no, size_t data_size, uint8_t *data_buffer)
+{
+	int ret;
+	int tx_count = 0;
+	struct pollfd fds = {
+		.fd     = sock,
+		.events = ZSOCK_POLLIN,
+	};
+
+	LOG_DBG("Client send data: block no %u, size %u", block_no, data_size);
+
+	/* Send data and poll for ACK response */
+	sys_put_be16(DATA_OPCODE, data_buffer);
+	sys_put_be16(block_no, data_buffer + 2);
+	do {
+		if (tx_count > TFTP_REQ_RETX) {
+			LOG_ERR("No more retransmits. Exiting");
+			return TFTPC_RETRIES_EXHAUSTED;
+		}
+		ret = send(sock, data_buffer, data_size + TFTP_HEADER_SIZE, 0);
+		if (ret < 0) {
+			LOG_ERR("send() error: %d", -errno);
+			return -errno;
+		}
+		tx_count++;
+	} while (poll(&fds, 1, CONFIG_TFTPC_REQUEST_TIMEOUT) <= 0);
+
+	/* Receive ACK from the TFTP Server. */
+	ret = recv(sock, tftpc_buffer, TFTPC_MAX_BUF_SIZE, 0);
+	if (ret < 0) {
+		LOG_ERR("recv() error: %d", -errno);
+		return -errno;
+	}
+
+	/* Check data send result */
+	if (ret >= TFTP_HEADER_SIZE) {
+		uint16_t opcode = sys_get_be16(tftpc_buffer);
+		uint16_t blockno = sys_get_be16(tftpc_buffer + 2);
+
+		LOG_DBG("Receive: opcode %u, block no %u, size %d", opcode, blockno, ret);
+
+		if (opcode != ACK_OPCODE || blockno != block_no) {
```

Comment:
I see a flaw in the current design - if the ACK is delayed for whatever reason, making us retransmit, but eventually arrives, the whole transmit produre will fail, as the routing will already expect the next block number. I think the routine should be more permissive here, i. e. allow receving ACKs with block number < that the current one, and simply ignore those.

Reply:
That might explains the problem I faced in upload streaming test.
Thanks for the suggestion!